[Profiler] tracking Optimizer (part 2 of Record Optimizer)

slgong-fb opened the pull request:
Summary:
Part 2 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- hooking from optimizer step
- PyOptCall Type
- declare data type for collection
- python binding
- simple unit test case

Test Plan: buck run mode/opt //caffe2/test:profiler

Differential Revision: D39402667

slgong-fb commented:
@albanD This PR is second part of the refactored #84063. I need an approval on this PR as you mentioned earlier about hook.
https://github.com/pytorch/pytorch/pull/84063#discussion_r956333848

robieta commented:
Discussion on instrumentation and migration to hooks if they are introduced to optimizer: https://github.com/pytorch/pytorch/pull/84063#discussion_r977370462

robieta reviewed (approved):
LGTM

slgong-fb committed:
[Profiler] tracking Optimizer (part 2 of Record Optimizer) (#84920)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/84920

Part 2 of Record Optimizer param_groups and states (https://github.com/pytorch/pytorch/pull/84063)
- hooking from optimizer step
- PyOptimizerCall Type
- declare data type for collection
- python binding
- simple unit test case

Test Plan: buck run mode/opt //caffe2/test:profiler

Reviewed By: robieta

Differential Revision: D39402667

fbshipit-source-id: 1dc10efbbf36fca5d822dd9c5aa7072ba0d9f581
